### PR TITLE
MAINT: Replace PyUString_ConcatAndDel in mapping.c.

### PR DESCRIPTION
**Pull request by charris**
This clarifies the exceptions raised in `mapping.c`. This work is broken out into a separate PR to simplify review.

### CODE REVIEW

File: numpy/core/src/multiarray/mapping.c [hunk 1]
```diff
@@ -2433,35 +2429,38 @@ mapiter_fill_info(PyArrayMapIterObject *mit, npy_index_info *indices,
 
     return 0;
 
-  broadcast_error:
+broadcast_error: ; // need semicolon to avoid compilation Cerror
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/mapping.c:2432]: And I thought I knew the C language... Perhaps
```suggestion
broadcast_error: ; // `;` is an empty statement, goto labels cannot precede declarations
```

File: numpy/core/src/multiarray/mapping.c [hunk 2]
```diff
@@ -2433,35 +2429,38 @@ mapiter_fill_info(PyArrayMapIterObject *mit, npy_index_info *indices,
...
             continue;
         }
-        tmp = convert_shape_to_string(
-                    PyArray_NDIM((PyArrayObject *)indices[i].object),
-                    PyArray_SHAPE((PyArrayObject *)indices[i].object),
-                    " ");
+
+        int ndim = PyArray_NDIM((PyArrayObject *)indices[i].object);
+        npy_intp *shape = PyArray_SHAPE((PyArrayObject *)indices[i].object);
+        PyObject * tmp = convert_shape_to_string(ndim, shape, " ");
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/mapping.c:2448]:
```suggestion
        PyObject *tmp = convert_shape_to_string(ndim, shape, " ");
```

**Review by eric-wieser — APPROVED** [review]
Looks good, only small nits. Thanks for teaching me some C I wish wasn't a thing.

File: numpy/core/src/multiarray/mapping.c (hunk 1 above)
Comment (charris) [reply numpy/core/src/multiarray/mapping.c:2432]: Took me a while to track that down. The error message was clear, but so weird that it took me a while to comprehend and track down the solution.

**Pull request merged**